08_irq_hanling/led_mod.c

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: GPL
+/*
+ * Copyright (c) 2021 Vasyl Yosypenko
+ */
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": %s: " fmt, __func__
+
+#include <linux/module.h>
+#include <linux/gpio.h>
+#include <linux/ktime.h>
+#include <linux/interrupt.h>
+
+#define GPIO_NUMBER(port, bit) (32 * (port) + (bit))
+
+
+/*  https://linux-sunxi.org/Xunlong_Orange_Pi_Zero#LEDs
+ *  Board config for OPI-Zero:
+ *  LED GREEN (PL10): GPIO_11_10
+ *  LED RED   (PA17): GPIO_0_17
+ *  BUTTON    (PG7) : GPIO_6_7
+ *
+ */
+
+#define LED_GREEN GPIO_NUMBER(11, 10)
+#define LED_RED GPIO_NUMBER(0, 17)
+#define BUTTON GPIO_NUMBER(6, 7)
+
+static int ledg_gpio = -1;
+static int ledr_gpio = -1;
+static int button_gpio = -1;
+static int button_state = -1;
+static int button_cnt = -1;
+
+static int irq;
+static int dev_id = 0;
+
+typedef enum {
+    ACTIVE,
+    NOT_ACTIVE
+}
+led_latch_state_t;
+
+static led_latch_state_t led_latch = ACTIVE;
+
+static irqreturn_t my_gpio_interrupt(int irq, void * devid)
+{
+   /*
+    * One button press:
+    * enable green LED, disable red LED
+    * press button one more time:
+    * disable green LED, enable red LED
+    *
+    */
+
+    if (ACTIVE == led_latch)
+    {
+        gpio_set_value(ledg_gpio, 1U);
+        gpio_set_value(ledr_gpio, 0U);
```

Comment:
This approach can produce very short pulses. Sometimes it is good idea to insert delay of use timers for create longer impulses.

Reply:
Thanks, I know. But in this code two LEDs are used, red - led**r**_gpio, and green one - led**g**_gpio